### PR TITLE
[Android] Support for PhoneCallHistory

### PR DESCRIPTION
**Pull request by pkar70**
GitHub Issue (If applicable): #3045

## PR Type

What kind of change does this PR introduce?
- Feature

## What is the current behavior?
 No PhoneCallHistory support

## What is the new behavior?
Support on Android plaftorm.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tested code with current [supported SDKs](https://github.com/unoplatform/uno/blob/master/README.md#uno-features)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [X] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [X] Contains **NO** breaking changes
- [X] Updated the [Release Notes](https://github.com/unoplatform/uno/tree/master/doc/ReleaseNotes)
- [ ] Associated with an issue (GitHub or internal)

## Other information
 Enums are defined for every platform, code is for Android only. As far as I know (from Internet digging), support for iOS would never happen, as it is impossible to access 'calllog' on iOS platform.
 Part of my app CallsStat, published on Amazon.

Internal Issue (If applicable):

Part of my app https://github.com/pkar70/CallStat

### CODE REVIEW

File: doc/ReleaseNotes/_ReleaseNotes.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Release notes
 
 ### Features
-
+- [Android] Support for PhoneCallHistory
```
Comment (jeromelaban) [inline doc/ReleaseNotes/_ReleaseNotes.md]:
```suggestion
- [Android] Support for `PhoneCallHistory`
```

File: src/Uno.UWP/Uno.csproj [hunk 2]
```diff
@@ -38,6 +38,24 @@
...
+	
+	<ItemGroup>
+	  <UpToDateCheckInput Remove="ApplicationModel\Calls\PhoneCallHistoryEntry.Android.cs" />
+	  <UpToDateCheckInput Remove="ApplicationModel\Calls\PhoneCallHistoryEntryAddress.Android.cs" />
+	  <UpToDateCheckInput Remove="ApplicationModel\Calls\PhoneCallHistoryEntryRawAddressKind.cs" />
+	  <UpToDateCheckInput Remove="ApplicationModel\Calls\PhoneCallHistoryEntryReader.Android.cs" />
+	  <UpToDateCheckInput Remove="ApplicationModel\Calls\PhoneCallHistoryManager.Android.cs" />
+	  <UpToDateCheckInput Remove="ApplicationModel\Calls\PhoneCallHistoryStore.Android.cs" />
+	  <UpToDateCheckInput Remove="ApplicationModel\Calls\PhoneCallHistoryStoreAccessType.cs" />
+	</ItemGroup>
```
Comment (jeromelaban) [inline src/Uno.UWP/Uno.csproj]: Please remove, VS has an issue that adds those lines that are not required

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryAddress.Android.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+﻿
+#if __ANDROID__
+namespace Windows.ApplicationModel.Calls
+{
+	public partial class PhoneCallHistoryEntryAddress
+	{
+		public PhoneCallHistoryEntryRawAddressKind RawAddressKind { get; set; }
+		public  string RawAddress { get; set; }
```
[the remark this answers is not in the document]
Comment (pkar70) [reply src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryAddress.Android.cs]: Spacing? Do you mean empty lines between properties (as RawAddressKind and RawAddress)?

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntry.Android.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+		public  TimeSpan? Duration { get; set; }
+
+		public  DateTimeOffset StartTime { get; set; }
+
+		public PhoneCallHistoryEntryAddress Address { get; set; }
+
+		public bool IsVoicemail { get; set; }
+		
+		public PhoneCallHistoryEntry()
```
[the remark this answers is not in the document]
Comment (pkar70) [reply src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntry.Android.cs]: Yes.
https://docs.microsoft.com/en-us/uwp/api/Windows.ApplicationModel.Calls.PhoneCallHistoryEntry

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryAddress.Android.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeromelaban) [reply src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryAddress.Android.cs]: It's about having consistent spacing between properties like the rest of the file.

**Comment by pkar70** [comment]
What test should do?
Besides, I tried to create one UI non-automatic test for CalendarDatePicker, failed (for reason uknown to me), and PR for CDP was marked as 'help needed'.
I know how to create automatic test, and such tests were included in many my Uno PR.

**Comment by pkar70** [comment]
Can build be restarted? "Publishing build artifacts failed with an error: Not found PathtoPublish" not related to changes in this PR.

**Comment by pkar70** [comment]
Can build be restarted? "Publishing build artifacts failed with an error: Not found PathtoPublish" not related to changes in this PR.

**Comment by pkar70** [comment]
"Removed method System.Void Windows.UI.Input.DraggingEventArgs..ctor() not found in ignore set.
Removed method System.Void Windows.Storage.FileIO..ctor() not found in ignore set." - errors not related to this PR. Can build be restarted?

**Comment by pkar70** [comment]
Can I get any review of this PR?

**Comment by pkar70** [comment]
This functionality would never be supported for iOS, as it is impossible to access 'calllog' on iOS platform. But, can this functionality be added to Uno for Android platform, i.e. could this PR be reviewed/accepted/merged?

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+			//	}
+			//	finally
+			//	{
+			//		current.RequestPermissionsResultWithResults -= handler;
+			//	}
+
+
+			//	return Windows.Devices.Geolocation.GeolocationAccessStatus.Denied;
+
+			//}
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:136]: This needs to be in a documentation file, you can take a look in the features directory.

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+		private static async Task<PhoneCallHistoryStore> RequestStoreAsyncTask(PhoneCallHistoryStoreAccessType accessType)
+		{
+			// UWP: AppEntriesReadWrite, AllEntriesLimitedReadWrite, AllEntriesReadWrite
+			// Android: Manifest has READ_CALL_LOG and WRITE_CALL_LOG, no difference between app/limited/full
+			// using: AllEntriesReadWrite as ReadWrite, and AllEntriesLimitedReadWrite as ReadOnly
+
+			if(accessType == PhoneCallHistoryStoreAccessType.AppEntriesReadWrite)
+			{   // should not happen, as this option is defined as NotImplemented in enum
+				throw new NotImplementedException();
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:27]: Use `NotSupportedException` for now, and specify an error message.

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+			}
+
+
+			var _histStore = new PhoneCallHistoryStore();
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:31]:
```suggestion
			var historyStore = new PhoneCallHistoryStore();

```

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+			{
+				return _histStore;
+			}
+
+			// since API 29, we should do something more:
+			// https://developer.android.com/reference/android/content/pm/PackageInstaller.SessionParams.html#setWhitelistedRestrictedPermissions(java.util.Set%3Cjava.lang.String%3E)
+
+			// do we have declared this permission in Manifest?
+			// it could be also Coarse, without GPS
+			Android.Content.Context context = Android.App.Application.Context;
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:44]: use `var` wherever possible.

```suggestion
			var context = Android.App.Application.Context;

```

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+					requestPermission.Add(Android.Manifest.Permission.WriteCallLog);
+				}
+			}
+
+			if (requestPermission.Count < 1)
+				return _histStore;
+
+			// in my tests, this ends with ERROR!! about null object reference inside TryGetPermission, on second call
+			foreach (var sPerm in requestPermission)
+				await Windows.Extensions.PermissionsHelper.TryGetPermission(CancellationToken.None, sPerm);
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:92]: use curly braces

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+				if (!await Windows.Extensions.PermissionsHelper.CheckPermission(CancellationToken.None, Android.Manifest.Permission.WriteCallLog))
+				{
+					requestPermission.Add(Android.Manifest.Permission.WriteCallLog);
+				}
+			}
+
+			if (requestPermission.Count < 1)
+				return _histStore;
+
+			// in my tests, this ends with ERROR!! about null object reference inside TryGetPermission, on second call
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:90]: This needs to be validatable somehow in a sample provided by this PR.

**Review by jeromelaban — COMMENTED** [review]
Add a sample in the sample app that shows the results of the invocations for the APIs you implemented.

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs (hunk 5 above)
Comment (pkar70) [reply src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:136]: info in `//` was outdated (some PRs ago).

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs [hunk 11]
```diff
@@ -0,0 +1,117 @@
+﻿#if __ANDROID__
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs:1]:
```suggestion
#nullable enable

﻿#if __ANDROID__

```

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs [hunk 12]
```diff
@@ -0,0 +1,102 @@
+﻿
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:1]:
```suggestion
#nullable enable
﻿
```

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+	// https://developer.samsung.com/galaxy/others/calllogs-in-android
+	public partial class PhoneCallHistoryEntryReader
+	{
+		// <uses-permission android:name="android.permission.READ_CONTACTS">  ? A nie calllog?
+
+		private Android.Database.ICursor _cursor = null;
+
+		public PhoneCallHistoryEntryReader()
+		{
+			Android.Content.ContentResolver cr = Android.App.Application.Context.ContentResolver;
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs:21]:
```suggestion
			var cr = Android.App.Application.Context.ContentResolver;

```

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+			{
+				_cursor.Close();
+				_cursor.Dispose();
+				_cursor = null;
+			}
+		}
+
+
+		public IAsyncOperation<IReadOnlyList<PhoneCallHistoryEntry>> ReadBatchAsync() => ReadBatchAsyncTask().AsAsyncOperation<IReadOnlyList<PhoneCallHistoryEntry>>();
+		private async Task<IReadOnlyList<PhoneCallHistoryEntry>> ReadBatchAsyncTask()
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs:47]:
```suggestion
		public IAsyncOperation<IReadOnlyList<PhoneCallHistoryEntry>> ReadBatchAsync() => ReadBatchAsyncTask().AsAsyncOperation<IReadOnlyList<PhoneCallHistoryEntry>>();
		
		private async Task<IReadOnlyList<PhoneCallHistoryEntry>> ReadBatchAsyncTask()

```

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+			var packageInfo =
+				context.PackageManager.GetPackageInfo(context.PackageName, Android.Content.PM.PackageInfoFlags.Permissions);
+			var requestedPermissions = packageInfo?.RequestedPermissions;
+			if (requestedPermissions is null)
+			{
+				throw new UnauthorizedAccessException("no permissions in Manifest defined (no permission at all)");
+			}
+
+			if(!Windows.Extensions.PermissionsHelper.IsDeclaredInManifest(Android.Manifest.Permission.ReadCallLog))
+				return null;
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:54]: Use braces for conditions

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+			if (!Windows.Extensions.PermissionsHelper.IsDeclaredInManifest(Android.Manifest.Permission.ReadContacts))
+				return null;
+
+			if (accessType == PhoneCallHistoryStoreAccessType.AllEntriesReadWrite)
+			{
+				if (!Windows.Extensions.PermissionsHelper.IsDeclaredInManifest(Android.Manifest.Permission.WriteCallLog))
+					return null;
+			}
+
+			List<string> requestPermission = new List<string>();
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:66]:
```suggestion
			List<string> requestPermission = new();

```

**Review by jeromelaban — COMMENTED** [review]
A few additional comments. We still need a sample in the samples app to validate the behavior.

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs (hunk 12 above)
Comment (pkar70) [reply src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:1]: It cannot be null, as RequestStoreAsyncTask returns PhoneCallHistoryStore not `PhoneCallHistoryStore?`, and sometimes it returns null.

**Comment by pkar70** [comment]
About sample...
Reading PhoneCallHistory is relevant only on Phones, and only after at least one call. Could it be done in SamplesApp?

Can it be sample outside Uno? https://github.com/pkar70/CallStat

**Comment by pkar70** [comment]
Will build restart by itself, or should it be restarted manually?

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs (hunk 16 above)
Comment (jeromelaban) [reply src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:81]: You'll likely need to rebase your PR.

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs (hunk 12 above)
Comment (pkar70) [reply src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:1]: ok

**Comment by jeromelaban** [comment]
> 

This is generally an error caused by visual studio itself. If the code compiles, you should be good.

File: src/SamplesApp/UITests.Shared/Windows_ApplicationModel/Calls/PhoneCallHistoryEntryReader.xaml.cs [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+using Windows.UI.Xaml.Controls.Primitives;
+using Windows.UI.Xaml.Data;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Media;
+using Windows.UI.Xaml.Navigation;
+using Windows.ApplicationModel.Calls;
+          
+namespace UITests.Shared.Windows_ApplicationModel.Calls
+{
+	[SampleControlInfo("Windows_ApplicationModel.Calls", "ContactReader")]
```
Comment (jeromelaban) [inline src/SamplesApp/UITests.Shared/Windows_ApplicationModel/Calls/PhoneCallHistoryEntryReader.xaml.cs:22]:
```suggestion
	[SampleControlInfo("Windows.ApplicationModel.Calls", "ContactReader")]
```

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Windows.Foundation;
+
+namespace Windows.ApplicationModel.Calls
+{
+	// https://developer.samsung.com/galaxy/others/calllogs-in-android
+	public partial class PhoneCallHistoryEntryReader
+	{
+		// <uses-permission android:name="android.permission.READ_CONTACTS">  ? A nie calllog?
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs:17]: Use English please.

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+
+		private Android.Database.ICursor? _cursor = null;
+
+		public PhoneCallHistoryEntryReader()
+		{
+			var cr = Android.App.Application.Context.ContentResolver;
+			if(cr is null)
+			{
+				throw new NullReferenceException("Windows.ApplicationModel.Calls.PhoneCallHistoryEntryReader.ctor, ContentResolver is null (impossible)");
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs:26]: Don't throw null reference, use `InvalidOperationException` instead.

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+			}
+
+			var oUri = Android.Provider.CallLog.Calls.ContentUri;
+			if (oUri is null)
+			{
+				throw new NullReferenceException("Windows.ApplicationModel.Calls.PhoneCallHistoryEntryReader.ctor, ContentUri is null (impossible)");
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs:32]:
```suggestion
				throw new InvalidOperationException("Windows.ApplicationModel.Calls.PhoneCallHistoryEntryReader.ctor, ContentUri is null (impossible)");

```

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+
+
+			_cursor = cr.Query(oUri,
+									null, 
+									null,
+									null,
+									Android.Provider.CallLog.Calls.DefaultSortOrder);   // == date DESC
+			if (_cursor is null)
+			{
+				throw new NullReferenceException("Windows.ApplicationModel.Calls.PhoneCallHistoryEntryReader.ctor, _cursor is null (impossible)");
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs:43]:
```suggestion
				throw new InvalidOperationException("Windows.ApplicationModel.Calls.PhoneCallHistoryEntryReader.ctor, _cursor is null (impossible)");

```

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+			}
+
+			// since API 29, we should do something more:
+			// https://developer.android.com/reference/android/content/pm/PackageInstaller.SessionParams.html#setWhitelistedRestrictedPermissions(java.util.Set%3Cjava.lang.String%3E)
+
+			var context = Android.App.Application.Context;
+			var packageManager = context.PackageManager;
+			if (packageManager is null)
+			{
+				throw new NullReferenceException("Windows.ApplicationModel.Calls.PhoneCallHistoryManager.RequestStoreAsyncTask, PackageManager is null (impossible)");
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:47]:
```suggestion
				throw new InvalidOperationException("Windows.ApplicationModel.Calls.PhoneCallHistoryManager.RequestStoreAsyncTask, PackageManager is null (impossible)");

```

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+			}
+
+			var packageName = context.PackageName;
+			if(packageName is null)
+			{
+				throw new NullReferenceException("Windows.ApplicationModel.Calls.PhoneCallHistoryManager.RequestStoreAsyncTask, PackageName is null (impossible)");
```
Comment (jeromelaban) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryManager.Android.cs:53]:
```suggestion
				throw new InvalidOperationException("Windows.ApplicationModel.Calls.PhoneCallHistoryManager.RequestStoreAsyncTask, PackageName is null (impossible)");

```

**Comment by jeromelaban** [comment]
I tried the app from the CI and this is what I get:
Screenshot_20220414-151935

Can you try it on your end ?

**Comment by pkar70** [comment]
can build be restarted?

File: src/SamplesApp/UITests.Shared/Windows_ApplicationModel/Calls/PhoneCallHistoryEntryReader.xaml.cs [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+			uiDuration.Text = "";
+
+			PhoneCallHistoryStore oCallHist;
+			try
+			{
+				oCallHist = await PhoneCallHistoryManager.RequestStoreAsync(PhoneCallHistoryStoreAccessType.AllEntriesLimitedReadWrite);
+			}
+			catch (Exception ex)
+			{
+				uiErrorMsg.Text = "Exception while RequestStoreAsync: " + ex.Message;
```
Comment (jeromelaban) [inline src/SamplesApp/UITests.Shared/Windows_ApplicationModel/Calls/PhoneCallHistoryEntryReader.xaml.cs:49]: I'm still getting a null reference exception here. Let's show the full trace:
```suggestion
				uiErrorMsg.Text = "Exception while RequestStoreAsync: " + ex;
```

File: src/SamplesApp/UITests.Shared/Windows_ApplicationModel/Calls/PhoneCallHistoryEntryReader.xaml [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+    xmlns:local="using:UITests.Shared.Windows_ApplicationModel.Calls"
+    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" 
+    mc:Ignorable="d">
+
+	<StackPanel Padding="20">
+		<Button Click="getCallLog_Click">Get call log</Button>
+		<TextBlock Text="" Foreground="Red" Name="uiErrorMsg"/>
```
Comment (jeromelaban) [inline src/SamplesApp/UITests.Shared/Windows_ApplicationModel/Calls/PhoneCallHistoryEntryReader.xaml:12]:
```suggestion
		<TextBlock Text="" Foreground="Red" Name="uiErrorMsg" TextWrapping="Wrap" />
```

**Comment by pkar70** [comment]
Removing 'draft' as there is sample added.

**Comment by agneszitte-nventive** [comment]
build re-started @pkar70

**Review by MartinZikmund — APPROVED** [review]

**Comment by MartinZikmund** [comment]
Did a small cleanup, a few fixes and expanded the sample to show the full log:

Screenshot_20220918-224015~2

**Comment by pkar70** [comment]
So, there is some chance that this code, used in my app available in Amazon store since May 13, 2020, will be finally included in official Uno distribution in September 2022 :)

File: src/SamplesApp/SamplesApp.Droid/Properties/AndroidManifest.xml [hunk 26]
```diff
@@ -2,6 +2,9 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android" package="uno.platform.unosampleapp" android:versionCode="1" android:versionName="1.0" android:installLocation="auto">
 	<uses-sdk android:minSdkVersion="19" android:targetSdkVersion="31" />
 	<!-- NOTE: Add permissions in AssemblyInfo.cs -->
```
Comment (Youssef1313) [inline src/SamplesApp/SamplesApp.Droid/Properties/AndroidManifest.xml:4]: @pkar70 @MartinZikmund Why are permissions added here while this comment mentions they should be in AssemblyInfo?

@jeromelaban This comment was added in https://github.com/unoplatform/uno/pull/3064, do you recall why permissions should be in AssemblyInfo not in AndroidManifest?

Either way, I guess we should just follow the comment and move the permissions.

File: src/Uno.UWP/Generated/3.0.0.0/Windows.ApplicationModel.Calls/PhoneCallHistoryStore.cs [hunk 27]
```diff
@@ -14,7 +14,7 @@ public  partial class PhoneCallHistoryStore
 			throw new global::System.NotImplementedException("The member IAsyncOperation<PhoneCallHistoryEntry> PhoneCallHistoryStore.GetEntryAsync(string callHistoryEntryId) is not implemented in Uno.");
 		}
 		#endif
-		#if __ANDROID__ || __IOS__ || NET461 || __WASM__ || __SKIA__ || __NETSTD_REFERENCE__ || __MACOS__
+		#if false || __IOS__ || NET461 || __WASM__ || __SKIA__ || __NETSTD_REFERENCE__ || __MACOS__
 		[global::Uno.NotImplemented("__ANDROID__", "__IOS__", "NET461", "__WASM__", "__SKIA__", "__NETSTD_REFERENCE__", "__MACOS__")]
```
Comment (Youssef1313) [inline src/Uno.UWP/Generated/3.0.0.0/Windows.ApplicationModel.Calls/PhoneCallHistoryStore.cs:18]: `__ANDROID__` must be removed in this line.

File: src/Uno.UWP/Generated/3.0.0.0/Windows.ApplicationModel.Calls/PhoneCallHistoryStore.cs [hunk 28]
```diff
@@ -2,7 +2,7 @@
 #pragma warning disable 114 // new keyword hiding
 namespace Windows.ApplicationModel.Calls
 {
-	#if __ANDROID__ || __IOS__ || NET461 || __WASM__ || __SKIA__ || __NETSTD_REFERENCE__ || __MACOS__
+	#if false || __IOS__ || NET461 || __WASM__ || __SKIA__ || __NETSTD_REFERENCE__ || __MACOS__
 	[global::Uno.NotImplemented]
```
Comment (Youssef1313) [inline src/Uno.UWP/Generated/3.0.0.0/Windows.ApplicationModel.Calls/PhoneCallHistoryStore.cs:6]: We should pass specific platforms to the attribute here.

File: src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs [hunk 29]
```diff
@@ -0,0 +1,140 @@
+﻿#nullable enable
+
+#if __ANDROID__
```
Comment (Youssef1313) [inline src/Uno.UWP/ApplicationModel/Calls/PhoneCallHistoryEntryReader.Android.cs:3]: I think this is not necessary. The file is already `.Android.cs` and is only included in build for Android.

File: src/SamplesApp/SamplesApp.Droid/Properties/AndroidManifest.xml (hunk 26 above)
Comment (jeromelaban) [reply src/SamplesApp/SamplesApp.Droid/Properties/AndroidManifest.xml:4]: It's generally easier to manipulate these flags in code, as they be placed pretty much anywhere, close to the code using them.

**Pull request merged**